prboom2/src/gl_preprocess.c

```diff
@@ -695,6 +705,163 @@ static void gld_PrecalculateSector(int num)
   Z_Free(lineadded);
 }
 
+static dboolean gld_IsRealLine(sector_t* sector, line_t* line)
+{
+  int i;
+  int v1c;
+  int v2c;
+
+  // Obvious case -- at least one side isn't sector itself
+  if (line->frontsector != sector || line->backsector != sector)
+    return true;
+
+  // Distinguish a normal mid-sector linedef from a mapping trick by
+  // checking connectedness of vertices
+  v1c = v2c = 0;
+  for (i = 0; i < sector->linecount; ++i)
+  {
+    if (sector->lines[i] == line)
+      continue;
+    if (sector->lines[i]->v1 == line->v1 || sector->lines[i]->v2 == line->v1)
+      v1c++;
+    if (sector->lines[i]->v2 == line->v2 || sector->lines[i]->v2 == line->v2)
+      v2c++;
+  }
+
+  // Line is still real if:
+  // - It's isolated (no neighbors), i.e. it's floating mid-sector
+  // - It has more than 1 neighbor per vertex, i.e. it divides the sector
+  return (v1c == 0 && v2c == 0) || (v1c > 1 && v2c > 1);
+}
+static inline dboolean gld_SeesRealSector(sector_t* sector, sector_t* other)
+{
+  return other != NULL && other != sector && (other->flags & SECTOR_IS_REAL);
+}
+
+static sector_t* gld_RealSectorAcrossLine(sector_t* sector, line_t* line)
+{
+  if (gld_SeesRealSector(sector, line->frontsector))
+    return line->frontsector;
+  if (gld_SeesRealSector(sector, line->backsector))
+    return line->backsector;
+  return NULL;
+}
+
+static void gld_MarkRealSectors(void)
+{
+  // Mark all sectors that are "real", i.e. that don't use self-referencing
+  // linedefs, or are transitively reachable from such a sector via sidedefs.
+  // This should be the vast majority of the map.
+  int i, j;
+  dboolean real;
+  dboolean changing;
+
+  // First, mark all sectors that are directly real (base case)
+  for (i = 0; i < numsectors; i++) {
+    real = true;
+    for (j = 0; j < sectors[i].linecount; j++) {
+      if (!gld_IsRealLine(&sectors[i], sectors[i].lines[j])) {
+        real = false;
+        break;
+      }
+    }
+    if (real)
+      sectors[i].flags |= SECTOR_IS_REAL;
+  }
+
+  // Now mark all transitively real sectors in passes (to avoid recursive calls)
```

Comment:
I'm concerned about a loop that may check tens of thousands of sectors repeatedly 😅

Reply:
The vast majority of sectors will be tagged by the first loop, meaning this loop will run at most once or twice in practice.  This does have worst case N^2 runtime, but it would only occur on a map which consists of N-1 self-referencing sectors chained end-to-end with sector N being real.  That's not a very realistic map.

This loop could be removed, but I was trying to be extremely paranoid about treating a sector as self-referencing when it isn't.

Reply:
I'll rework this to use an explicit queue and avoid visiting unnecessary sectors.